### PR TITLE
Set local pipeline thread stack size

### PR DESCRIPTION
**Pull request by iSazonov**
## PR Summary
Related #4268.

1. Set local pipeline thread stack size to 10MB by default.
2. Read the `PipelineMaxStackSizeMB` configuration value from config.

Thread stack OS system default size is:
- on Windows - 1MB
- on MacOs - 0.5MB
- on Linux - 8MB

We should use the same stack size for all platforms to get predictable behavior.
This value should match the Windows PowerShell value for backward compatibility (I expect that .Net Core and Windows PowerShell consume the stack as well as .Net Framework and Windows PowerShell).
Perhaps default 10MB may be too large if we use many local pipelines for paralleling. We can reconsider this later.

## PR Checklist

Note: Please mark anything not applicable to this PR `NA`.

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [x] Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [NA] User facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [NA] Issue filed - Issue link:
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [NA] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
    - [NA] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.

### CODE REVIEW

File: src/System.Management.Automation/engine/PSConfiguration.cs [hunk 1]
```diff
@@ -548,6 +548,7 @@ internal sealed class PowerShellPolicies
         public ProtectedEventLogging ProtectedEventLogging { get; set; }
         public Transcription Transcription { get; set; }
         public UpdatableHelp UpdatableHelp { get; set; }
+        public UpdatableHelp PipelineMaxStackSizeMB { get; set; }
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/PSConfiguration.cs]: The `PipelineMaxStackSizeMB` setting is not a policy in windows powershell, and thus it shouldn't be added to `PowerShellPolicies`. It should be a setting like `psmodulepath`, or `executionpolicy` or `ConsolePrompting`.

THe `PipelineMaxStackSizeMB` was stored at `SOFTWARE\Microsoft\PowerShell\1\ShellIds` in windows powershell, not at `SOFTWARE\Policies\...`

File: src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs [hunk 2]
```diff
@@ -247,54 +248,35 @@ private void SetupInvokeThread(Thread invokeThread, bool changeName)
...
-            {
-                int i = ReadRegistryInt("PipelineMaxStackSizeMB", 10);
-                if (i < 10)
-                    i = 10; // minimum 10MB
-                else if (i > 100)
-                    i = 100; // maximum 100MB
-                return i * 1000000;
-            }
-        }
+        internal static int StackSize => GetPipelineMaxStackSizeOption(10);
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs]: If we don't cache the stack size, then this static property seems not necessary as it only adds one more redirection to call `GetPipelineMaxStackSizeOption`.

**Review by daxian-dbw — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/engine/PSConfiguration.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/PSConfiguration.cs]: I believe [previously](https://github.com/PowerShell/PowerShell/issues/5809#issuecomment-356059863) we agreed that PowerShell Core is a separate product and it should have its policies and not use policies and configs of Windows PowerShell.
Why don't we have consistency between the settings in the file and the registry if we the consistency by default on Unix? I wouldn't be surprised if MSFT decides to port Registry API to make it easier to port applications. (I would take this into account).
Also if the settings are in different places, it is difficult to find it and makes it very difficult to configure.
Another thought is that we must put (later?) config options and  GPO policies in different registry keys.

I think we should develop this my [comment](https://github.com/PowerShell/PowerShell/issues/5809#issuecomment-355843863) about priority - maybe open new tracking Issue?
Also we could open new tracking Issue (?) to enumerate config options and to solve where put different options - in policy key or/and in config key.

Today I suggest:
1. Have `PipelineMaxStackSizeMB` (and most other options) in registry and file.
2. Put `PipelineMaxStackSizeMB` in "Software\Policies\Microsoft\PowerShellCore". Later we can move/duplicate it in "Software\Microsoft\PowerShellCore" - today we haven't code to process the registry branches in priority.

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/PSConfiguration.cs]: I didn't mean `PipelineMaxStackSizeMB` should be put in Registry on Windows. I think this setting should be in the config file only on both Windows and Unix.

As far as I understand, only the GPO related settings will live in Registry and `powershell.config.json` on Windows for PowerShell Core (registry setting takes precedence). Non-GPO settings that used to be stored in Registry in Windows PowerShell now should live in the config file only in PowerShell Core, such as `executionpolicy`, `ConsolePrompting` and many others. You can take a look at `PSConfiguration.cs` to find more of them.

The type `PowerShellPolicies` only represents the GPO settings. `PipelineMaxStackSizeMB` is not a policy setting but a normal powershell setting, so it shouldn't be put in `PowerShellPolicies`.

Comment (iSazonov) [reply src/System.Management.Automation/engine/PSConfiguration.cs]: @daxian-dbw I wanted to create Issue but the text was too big and I made a document #5947

**Comment by iSazonov** [comment]
@daxian-dbw What is "policy" in the context? 

`Software\Policies\Microsoft\PowerShellCore` - I believe it is "policy"?

`Software\Microsoft\PowerShell\1\ShellIds` or 
`Software\PowerShellCore`  - I believe it is "regular settings"? Do you say that we shouldn't use Registry for this settings and keep its only in a config file?

**Comment by daxian-dbw** [comment]
> What is "policy" in the context?

It's short for "Group Police Settings". Those are the settings that are originally under "Software\Policies". Each of the setting has a corresponding GPO in "Local Group Policy Editor" in Control Panel. The new PowerShell Core ones will also be available in the editor. We will publish a `.admx` file for them.

The settings that are not originally under "Software\Policies" are regular powershell settings. As far as I understand, we don't use Registry for those settings but only refer to the config file.

**Comment by iSazonov** [comment]
Thanks for clarify!
>As far as I understand, we don't use Registry for those settings but only refer to the config file.

Is this PowerShell MSFT team solution? And why? 

Registry is common place for Windows applications to store settings:
https://msdn.microsoft.com/en-us/library/windows/desktop/ms724871(v=vs.85).aspx
https://msdn.microsoft.com/en-us/library/ms973902.aspx

This is critical for enterprises. By using GPO/GPP, they can centrally manage (distribute polices and settings too) applications on workstations and server farms.
For example, if they receive feedback from employees that their scripts fail because of a stack overflow, the enterprise administrator can easily propagate the parameter value as a default that does not overlap local settings at the application or user level.

**Comment by SteveL-MSFT** [comment]
For PSCore6, we want to have configuration not in the registry.  This makes it easier to deploy via zip, for example, as well as support side-by-side more simply.  I do agree that for an enterprise, we should support GPO.  In which case, there is no issue with having both configuration and policy.

**Comment by iSazonov** [comment]
>This makes it easier to deploy via zip, for example, as well as support side-by-side more simply.

The ability to have settings in the registry does not conflict with zip and side-by-side. (We could even enhance our API to support applications by type and name)

@SteveL-MSFT What your conclusion about #5947? - I'd move it in RFC repo but it's frozen. I would have moved RFC repo here with label like `RFC-*` - so we get more feedback.

**Comment by iSazonov** [comment]
I created https://github.com/PowerShell/PowerShell-RFC/pull/111.

Should we wait the RFC?

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this and don't believe that the stack size should even be a configurable setting until there is a clear need to make it settable.  Recommendation is to not even have it settable despite what Windows PowerShell did.

**Comment by iSazonov** [comment]
@SteveL-MSFT Stack size default is different on different platforms - should we set the same default _internally_ to get predictable behavior on all platforms? I think it make sense.

**Comment by SteveL-MSFT** [comment]
@iSazonov yes, @PowerShell/powershell-committee agrees to changing the default on all systems.  We thought the only point that needed discussion from the committee was configuration.

**Comment by daxian-dbw** [comment]
A new PR was submitted based on the committee's review: #6224 
I will close this PR.

**Pull request closed without merging**